kof-operator/internal/controller/configmap_controller.go

```diff
@@ -0,0 +1,576 @@
+		}
+		maps.Copy(oldRule.Labels, newRule.Labels)
+	}
+	if newRule.Annotations != nil {
+		if oldRule.Annotations == nil {
+			oldRule.Annotations = make(map[string]string, len(newRule.Annotations))
+		}
+		maps.Copy(oldRule.Annotations, newRule.Annotations)
+	}
+}
+
+// Get a map with YAML files from the nested map.
+func getAlertFiles(
+	ctx context.Context,
+	clusterGroupAlertRules map[string]map[string]AlertRules,
+) (map[string]string, error) {
+	log := log.FromContext(ctx)
+	files := map[string]string{}
+
+	for clusterName, groupRules := range clusterGroupAlertRules {
+		for groupName, rules := range groupRules {
+			fileName := groupName + ".yaml"
+			if clusterName != DefaultClusterName {
+				fileName = fmt.Sprintf("__%s__%s", clusterName, fileName)
+			}
+
+			rulesSlice := slices.Collect(maps.Values(rules))
+			slices.SortFunc(rulesSlice, func(a, b promv1.Rule) int {
+				return strings.Compare(a.Alert, b.Alert)
+			})
+
+			for _, rule := range rulesSlice {
+				if rule.Labels == nil {
+					rule.Labels = make(map[string]string)
+				}
+				rule.Labels["alertgroup"] = groupName
+				// If we find that adding `{cluster="cluster1"}` to `.Values.clusterRulesPatch`
+				// and `{cluster!~"^cluster1$|^cluster10$"}` to `.Values.defaultRulesPatch`
+				// manually is a problem, we can update `rule.Expr` automatically here with
+				// https://github.com/prometheus/prometheus/blob/main/promql/parser/ast.go
+			}
+
+			prometheusRuleSpec := promv1.PrometheusRuleSpec{
+				Groups: []promv1.RuleGroup{
+					{Name: groupName, Rules: rulesSlice},
+				},
+			}
+
+			yamlBytes, err := yaml.Marshal(prometheusRuleSpec)
+			if err != nil {
+				log.Error(
+					err, "failed to marshal alert rules",
+					"cluster", clusterName,
+					"group", groupName,
+				)
+				return nil, err
+			}
+			files[fileName] += string(yamlBytes)
+		}
+	}
+
+	return files, nil
+}
+
+// Update `ConfigMap` generated by `kof-operator` with the given `data`.
+// Either `configMap` or `namespace` and `name` should be provided.
+func (r *ConfigMapReconciler) updateConfigMap(
+	ctx context.Context,
+	configMap *corev1.ConfigMap,
+	namespace string,
+	name string,
+	data map[string]string,
+) error {
+	log := log.FromContext(ctx)
+
+	if configMap == nil {
+		configMap = &corev1.ConfigMap{}
+		namespacedName := types.NamespacedName{Namespace: namespace, Name: name}
+		if err := r.Get(ctx, namespacedName, configMap); err != nil {
+			log.Error(err, "failed to get ConfigMap",
+				"configMap", namespacedName,
+			)
+			return err
+		}
+	}
+
+	namespacedName := types.NamespacedName{
+		Namespace: configMap.Namespace,
+		Name:      configMap.Name,
+	}
+
+	if maps.Equal(configMap.Data, data) {
+		utils.LogEvent(
+			ctx,
+			"ConfigMapUpdateNotNeeded",
+			"No need to update ConfigMap",
+			configMap,
+			nil,
+			"configMap", namespacedName,
+		)
+		return nil
+	}
+
+	if configMap.Labels[utils.KofGeneratedLabel] != "true" {
+		utils.LogEvent(
```

Comment:
Events totally makes sense when an action failed or succeeded, but if there is not action taken, probably we should only log, but don't emit an event to avoid too much noise.